src/Jackett.Common/Indexers/CardigannIndexer.cs

```diff
@@ -1463,39 +1472,65 @@ protected override async Task<IEnumerable<ReleaseInfo>> PerformQuery(TorznabQuer
                 {
                     try
                     {
-                        var SearchResultParser = new HtmlParser();
-                        var SearchResultDocument = SearchResultParser.ParseDocument(results);
-
-                        // check if we need to login again
-                        var loginNeeded = CheckIfLoginIsNeeded(response, SearchResultDocument);
-                        if (loginNeeded)
                         {
-                            logger.Info(string.Format("CardigannIndexer ({0}): Relogin required", Id));
-                            var LoginResult = await DoLogin();
-                            if (!LoginResult)
-                                throw new Exception(string.Format("Relogin failed"));
-                            await TestLogin();
-                            response = await RequestWithCookiesAsync(searchUrl, method: method, data: queryCollection);
-                            if (response.IsRedirect && SearchPath.Followredirect)
-                                await FollowIfRedirect(response);
-
-                            results = response.ContentString;
-                            SearchResultDocument = SearchResultParser.ParseDocument(results);
+                            var SearchResultParser = new HtmlParser();
+                            var SearchResultDocument = SearchResultParser.ParseDocument(results);
+
+                            // check if we need to login again
+                            var loginNeeded = CheckIfLoginIsNeeded(response);
+                            if (loginNeeded)
```

Comment:
I not completely certain, but I don't think the login-again is working.
I've had two indexers which were returning 0 results errors, but when I use the wrench and saved the TEST then worked ok.

Reply:
Ah, I saw this as well. I did see some successful login-again results too though, so I put it down to being because I had transferred the saved indexer configs over from my Docker install to the Windows sandbox.

I've changed `response` to `SearchResultDocument` to see if that makes a difference.

Reply:
TL;DR - it's broken, but only slightly ;P

So I compared v0.20.384 (this build from before the last change) and v0.20.371.

Out of all but 4 of the semi-private indexers, 25 needed to relogin for both, and 2 fail for both (both expected). 2 fail for this build that don't fail for v0.20.371 - riperam and sktorrent-org.

Out of all the private trackers I have, 26 need to relogin for both, and 2 fail for both (1 expected, 1 I'm not sure why it fails for both - seedfile). 4 fail for this build that don't fail for v0.20.371 - anasch.cc, arabp2p, mircrew, rptorrents.

For the ones that fail just for this build, there's no attempt to relogin made, it performs a search, finds no results (because they get a mix of redirects and 'you are not signed in' pages), and it just reports the error.

Reply:
This block is trying to parse with HTMLParser no matter the response type.

```
var SearchResultParser = new HtmlParser();
var SearchResultDocument = SearchResultParser.ParseDocument(results);
```
Can we just move the login-again logic down into the Html path below? This kills re-login for xml for now, but should result in all HTML indexers working as before. Also if we do that, we can revert the changes to `CheckIfLoginIsNeeded` above

Also, seems to be wrapped with extra set of brackets which are not needed.

Reply:
This block should work

```
                    try
                    {
                        IHtmlCollection<IElement> rowsDom;

                        if (SearchPath.Response != null && SearchPath.Response.Type.Equals("xml"))
                        {
                            var SearchResultParser = new XmlParser();
                            var SearchResultDocument = SearchResultParser.ParseDocument(results);

                            if (Search.Preprocessingfilters != null)
                            {
                                results = applyFilters(results, Search.Preprocessingfilters, variables);
                                SearchResultDocument = SearchResultParser.ParseDocument(results);
                                logger.Debug(string.Format("CardigannIndexer ({0}): result after preprocessingfilters: {1}", Definition.Id, results));
                            }

                            var rowsSelector = applyGoTemplateText(Search.Rows.Selector, variables);
                            rowsDom = SearchResultDocument.QuerySelectorAll(rowsSelector);
                        }
                        else
                        {
                            var SearchResultParser = new HtmlParser();
                            var SearchResultDocument = SearchResultParser.ParseDocument(results);

                            // check if we need to login again
                            var loginNeeded = CheckIfLoginIsNeeded(response, SearchResultDocument);
                            if (loginNeeded)
                            {
                                logger.Info(string.Format("CardigannIndexer ({0}): Relogin required", Id));
                                var LoginResult = await DoLogin();
                                if (!LoginResult)
                                    throw new Exception(string.Format("Relogin failed"));
                                await TestLogin();
                                response = await RequestWithCookiesAsync(searchUrl, method: method, data: queryCollection);
                                if (response.IsRedirect && SearchPath.Followredirect)
                                    await FollowIfRedirect(response);

                                results = response.ContentString;
                                SearchResultDocument = SearchResultParser.ParseDocument(results);
                            }

                            checkForError(response, Definition.Search.Error);

                            if (Search.Preprocessingfilters != null)
                            {
                                results = applyFilters(results, Search.Preprocessingfilters, variables);
                                SearchResultDocument = SearchResultParser.ParseDocument(results);
                                logger.Debug(string.Format("CardigannIndexer ({0}): result after preprocessingfilters: {1}", Id, results));
                            }

                            var rowsSelector = applyGoTemplateText(Search.Rows.Selector, variables);
                            rowsDom = SearchResultDocument.QuerySelectorAll(rowsSelector);

                        }

                        var Rows = new List<IElement>();
                        foreach (var RowDom in rowsDom)
                        {
                            Rows.Add(RowDom);
                        }
```

Reply:
Same indexers failing to relogin :/

Reply:
hmm odd, this block should treat it no different than existing if it's not xml... It's literally the same code, just stuffed in an else

Did you revert the checkForError and CheckIfLoginIsNeeded back to original?

Reply:
Only used the above block, at which point it errored out again with `error CS1501: No overload for method 'CheckIfLoginIsNeeded' takes 2 arguments`, so I removed `SearchResultDocument`.

Reply:
That means your CheckLoginNeeded method is still modified as compared to master. Revert that method back to match master and ditch all the changes we did earlier

Reply:
Yep, that did it, thanks again!